MIPs/mip-3.md

```diff
@@ -0,0 +1,184 @@
+
+This proposal aims to solve a key pain point for dApp developers and users - the inability to easily swap tokens within the dApp interface.
+
+Currently, if a dApp requires users to swap from token A to token B to perform a specific operation, the user has to exit the dApp, go to MetaMask or third parties, search for the token the user needs and execute the swap, then return to the dApp. This disjointed workflow results in a poor user experience.
+
+Implementing `wallet_swapAsset` would allow the entire token swap process to occur seamlessly within the dApp without requiring external integration. This provides the following benefits:
+
+- Improved usability - Users can get the right tokens for dApp operations in fewer steps.
+- Better dApp integration - dApps can build and control the swap experience natively with just a few lines.
+- Increased adoption - More intuitive swap workflow lowers barrier to dApp usage.
+- Interoperability - Standard method works across different wallets.
+- Increase revenue - Easier adoption means more users will use MetaMask swaps.
+
+Comparing with other alternative, implementing it using swaps, would create more friction and less adoption since the users would need to install that specific snap to be able to perform a swap action. The all goal is to reduce friction to swaps.
+
+Overall, this proposal fills a clear gap in dApp capabilities related to swapping tokens. Implementing it would provide significant value to both developers and users in the Ethereum ecosystem.
+
+# Usage Example
+
+```
+ethereum.request({
+  method: 'wallet_swapAsset',
+  params: [{
+    from: [{
+      token_address: '0x1234567890abcdefABCDEF1234567890ABCDEF',
+      amount: '0xDE0B6B3A7640000',
+    }],
+    to: {
+      token_address: '0xabcdef1234567890ABCDEF1234567890abcdef',
+    },
+    user_address: '0x0000000000000000000000000000000000000000'
+  },
+  ]
+});
+
+```
+
+## Proposal Specification
+
+The new JSON RPC method wallet_swapAsset should be implemented with the following parameters:
+
+- `from`: An object containing details about the source token. It should include:
+
+  - `token_address`: The address of the source token.
+  - `amount`: The amount on wei and hexadecimal format of source token to be swapped.
+
+- `to`: An object containing details about the destination token. It should include:
+
+  - `token_address`: The address of the destination token.
+
+- `user_address`: An string containing the address connected to the dapp.
+
+- `referall_code`: (Optional) Future implementation will allow adding a referall code to bigger adoption of the json rpc method across the builders community
+
+- `send_to`: (Optional) Allows integration of the implementation of the swap send to feature.
+
+MetaMask will interpret the method call and perform the necessary validations and operations to initiate the token swap.
+
+## Caveats
+
+The implementation of `wallet_swapAsset` will be new and this could have some time for adoption.
```

Comment:
```suggestion
The implementation of `wallet_swapAsset` will be new and may take some time to be adopted.
```

Reply:
I'm not sure this caveat is necessary. Ramp up time for adoption can be assumed.

Reply:
Yeah, I aggree, removed!